deepmd/pt/model/model/dp_spin_model.py

```diff
@@ -0,0 +1,372 @@
+
+    @staticmethod
+    def extend_nlist(extended_atype, nlist):
+        nframes, nloc, nnei = nlist.shape
+        nall = extended_atype.shape[1]
+        nlist_mask = nlist != -1
+        nlist[nlist == -1] = 0
+        nlist_shift = nlist + nall
+        nlist[~nlist_mask] = -1
+        nlist_shift[~nlist_mask] = -1
+        self_spin = torch.arange(0, nloc, dtype=nlist.dtype, device=nlist.device) + nall
+        self_spin = self_spin.view(1, -1, 1).expand(nframes, -1, -1)
+        # self spin + real neighbor + virtual neighbor
+        # nf x nloc x (1 + nnei + nnei)
+        extended_nlist = torch.cat([self_spin, nlist, nlist_shift], dim=-1)
+        # nf x (nloc + nloc) x (1 + nnei + nnei)
+        extended_nlist = torch.cat(
+            [extended_nlist, -1 * torch.ones_like(extended_nlist)], dim=-2
+        )
+        return extended_nlist
+
+    @staticmethod
+    def extend_mapping(mapping, nloc: int):
+        return torch.cat([mapping, mapping + nloc], dim=-1)
+
+    @staticmethod
+    def switch_virtual_loc(extended_tensor, nloc: int):
+        """
+        Switch the virtual atoms of nloc ones from [nall: nall+nloc] to [nloc: nloc+nloc],
+        to assure the atom types of first nloc * 2 atoms in nall * 2 to be right.
+        """
+        nframes, nall_double = extended_tensor.shape[:2]
+        nall = nall_double // 2
+        swithed_tensor = torch.zeros_like(extended_tensor)
+        swithed_tensor[:, :nloc] = extended_tensor[:, :nloc]
+        swithed_tensor[:, nloc : nloc + nloc] = extended_tensor[:, nall : nall + nloc]
+        swithed_tensor[:, nloc + nloc : nloc + nall] = extended_tensor[:, nloc:nall]
+        swithed_tensor[:, nloc + nall :] = extended_tensor[:, nloc + nall :]
+        return swithed_tensor
+
+    @staticmethod
+    def switch_nlist(nlist_updated, nall: int):
+        nframes, nloc_double = nlist_updated.shape[:2]
+        nloc = nloc_double // 2
+        first_part_index = (nloc <= nlist_updated) & (nlist_updated < nall)
+        second_part_index = (nall <= nlist_updated) & (nlist_updated < (nall + nloc))
+        nlist_updated[first_part_index] += nloc
+        nlist_updated[second_part_index] -= nall - nloc
+        return nlist_updated
+
+    def extend_switch_input(
+        self,
+        extended_coord,
+        extended_atype,
+        extended_spin,
+        nlist,
+        mapping: Optional[torch.Tensor] = None,
+    ):
+        """
+        Add `extended_spin` into `extended_coord` to generate virtual atoms, and extend `nlist` and `mapping`.
+        Note that the final `extended_coord_updated` with shape [nframes, nall + nall, 3] has the following order:
+        - [:, :nloc]: original nloc real atoms.
+        - [:, nloc: nloc + nloc]: virtual atoms corresponding to nloc real atoms.
+        - [:, nloc + nloc: nloc + nall]: ghost real atoms.
+        - [:, nloc + nall: nall + nall]: virtual atoms corresponding to ghost real atoms.
+        """
```

Comment:
this is incorrect. the ghost atoms should always put behind the local atoms. 
I guess the forces should still be correct if you were use a wrong convention, but the virial calculation would fail.